### PR TITLE
[SYCLomatic] Emit warning for shared local memory if the data type defined by using or typedef

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/test/dpct/kernel_without_name.cu [hunk 1]
```diff
@@ -277,3 +277,18 @@ void run_foo7(T *a, const T *b, const T *c, const T *d, const T *e, const int f,
     foo_kernel6<<<grid, block, 0, stream>>>(a, b, c, d, e, f, g, h);
   }
 }
+
+#ifndef BUILD_TEST
+template <typename T> struct kernel_type_t {
+  using Type = T;
+};
+
+template <typename T> __global__ void foo_kernel7() {
```
Comment (zhiweij1) [inline clang/test/dpct/kernel_without_name.cu]: add check line for 286

File: clang/test/dpct/kernel_without_name.cu [hunk 2]
```diff
@@ -277,3 +277,18 @@ void run_foo7(T *a, const T *b, const T *c, const T *d, const T *e, const int f,
...
+  //CHECK:/*
+  //CHECK-NEXT:DPCT1124:{{[0-9]+}}: This type is used in other places. You may need adjust the type definition location.
+  //CHECK-NEXT:*/
+  //CHECK-NEXT:using Tk = typename kernel_type_t<T>::Type;
+  using Tk = typename kernel_type_t<T>::Type;
+  __shared__ Tk mem[256];
+}
+#endif
```
Comment (zhiweij1) [inline clang/test/dpct/kernel_without_name.cu:334]: add kernel call

File: clang/test/dpct/kernel_without_name.cu [hunk 3]
```diff
@@ -277,3 +277,18 @@ void run_foo7(T *a, const T *b, const T *c, const T *d, const T *e, const int f,
...
 }
+
+#ifndef BUILD_TEST
+template <typename T> struct kernel_type_t {
+  using Type = T;
+};
+
+template <typename T> __global__ void foo_kernel7() {
+  //CHECK:/*
+  //CHECK-NEXT:DPCT1124:{{[0-9]+}}: This type is used in other places. You may need adjust the type definition location.
```
Comment (zhiweij1) [inline clang/test/dpct/kernel_without_name.cu]: other places ==> call path from paralell_for sycl kernel call to current function ...

File: clang/test/dpct/kernel_without_name.cu [hunk 4]
```diff
@@ -277,3 +277,34 @@ void run_foo7(T *a, const T *b, const T *c, const T *d, const T *e, const int f,
...
   }
 }
+
+#ifndef BUILD_TEST
+template <typename T> struct kernel_type_t {
+  using Type = T;
+};
+
+//CHECK:template <typename T> void foo_kernel7(Tk *mem) {
+//CHECK-NEXT:  /*
```
Comment (zhiweij1) [inline clang/test/dpct/kernel_without_name.cu]: Move warning to Line 286 and kernel call location, and tell user the Tk is defined in function foo_kernel7

File: clang/test/dpct/kernel_without_name.cu [hunk 5]
```diff
@@ -277,3 +277,58 @@ void run_foo7(T *a, const T *b, const T *c, const T *d, const T *e, const int f,
...
+
+#ifndef BUILD_TEST
+template <typename T> struct kernel_type_t {
+  using Type = T;
+};
+
+// CHECK: template <typename T> 
+// CHECK-NEXT: void foo_device7(int a,
+// CHECK-NEXT:                  int b/*
+// CHECK-NEXT:                  DPCT1124:{{[0-9]+}}: The type "Tk" defined in function "foo_device7" is used as the parameter type in all functions in the call path from the corresponding sycl::handler::parallel_for() to the current function. You may need to adjust the type definition location.
```
Comment (zhimingwang36) [inline clang/test/dpct/kernel_without_name.cu]: suggest move the DPCT warning to L285.

Comment (zhiweij1) [reply clang/test/dpct/kernel_without_name.cu]: done

File: clang/test/dpct/kernel_without_name.cu [hunk 6]
```diff
@@ -277,3 +277,58 @@ void run_foo7(T *a, const T *b, const T *c, const T *d, const T *e, const int f,
...
+void foo_kernel7(int a,
+                 int b) {
+  foo_device7<T>(a, b);
+}
+
+template <typename T> 
+void run_foo8() {
+  // CHECK: int i;
+  // CHECK-NEXT: /*
+  // CHECK-NEXT: DPCT1124:{{[0-9]+}}: The type "Tk" defined in function "foo_device7" is used as the parameter type in all functions in the call path from the corresponding sycl::handler::parallel_for() to the current function. You may need to adjust the type definition location.
```
Comment (zhimingwang36) [inline clang/test/dpct/kernel_without_name.cu]:
```suggestion
  // CHECK-NEXT: DPCT1124:{{[0-9]+}}: The type "Tk" defined in function "foo_device7" is used as the parameter type in all functions in the call path from the sycl::handler::parallel_for() to the "foo_device7" function. You may need to adjust the type definition location.
```

Comment (zhiweij1) [reply clang/test/dpct/kernel_without_name.cu]: fixed.

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**